File: python/42_Trapping_Rain_Water.py
#!/usr/bin/python
import sys
import os

class Solution(object):
    def trap(self, height):
    	n, vol = len(height), 0
    	leftMax, rightMax = [0] * n, [0] * n
    	for i in range(1, n):
    		leftMax[i] = max(leftMax[i - 1], height[i - 1])

    	for i in range(n - 2, 0, -1):
    		rightMax[i] = max(rightMax[i + 1], height[i + 1])
    		boundary = min(leftMax[i], rightMax[i])
    		if boundary > height[i]:
    			vol += boundary - height[i]
        return vol
        


def main():
    s = Solution()
    print s.trap([0,1,0,2,1,0,1,3,2,1,2,1])

if __name__ == "__main__":
    main()

File: c++/42_Trapping_Rain_Water.cpp
#include <iostream>
#include <vector>
#include <map>
#include <string>

using namespace std;


class Solution {
public:
    int trap(vector<int>& height) {
    	int n = height.size(), vol = 0;
    	vector<int> leftMax(n, 0);
    	vector<int> rightMax(n, 0);
    	for (int i = n - 2; i > 0; i--)
    		rightMax[i] = max(rightMax[i + 1], height[i + 1]);

    	for (int i = 1; i < n; i++) {
    		leftMax[i] = max(leftMax[i - 1], height[i - 1]);
    		int boundary = min(leftMax[i], rightMax[i]);
    		if (boundary > height[i])
    			vol += boundary - height[i];
    	}
    	return vol;
    }
};

